rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py

```diff
@@ -341,72 +362,225 @@ def _convert_attribute_tokens_to_texts(
+        if len(vocabulary) > len(existing_vocabulary):
             logger.warning(
-                "Unable to train a shared CountVectorizer. "
-                "Leaving an untrained CountVectorizer"
+                f"New data contains vocabulary of size {len(vocabulary)} for attribute {attribute} "
+                f"which is larger than the maximum vocabulary size({len(existing_vocabulary)}) "
+                f"of the original model. Some tokens will have to be dropped "
+                f"in order to continue training. It is advised to re-train the "
+                f"model from scratch on complete data."
             )
+        for token in vocabulary:
+            if token not in existing_vocabulary:
+                existing_vocabulary[token] = available_empty_index
+                del existing_vocabulary[f"{BUFFER_SLOTS_PREFIX}{available_empty_index}"]
+                available_empty_index += 1
+                if available_empty_index == len(existing_vocabulary):
+                    break
+        self._set_vocabulary(attribute, existing_vocabulary)
+
+    def _get_additional_vocabulary_size(
+        self, attribute: Text, existing_vocabulary_size: int
+    ) -> int:
+        """Get additional vocabulary size to be saved for incremental training.
+
+        If `self.additional_vocabulary_size` is not None,
+        we return that as the user should have specified
+        this number. If not then we take the default
+        additional vocabulary size which is 1/2 of the
+        current vocabulary size.
+        Args:
+            attribute: Message attribute for which additional vocabulary size should be computed.
+            existing_vocabulary_size: Current size of vocabulary learnt from the training data.
+
+        Returns:
+            Size of additional vocabulary that should be set aside for incremental training.
+        """
+        # Vocabulary expansion for INTENTS, ACTION_NAME
+        # and INTENT_RESPONSE_KEY is currently not supported.
+        if attribute not in DENSE_FEATURIZABLE_ATTRIBUTES:
+            return 0
+        if self.additional_vocabulary_size.get(attribute, None) is not None:
+            return self.additional_vocabulary_size[attribute]
+        return int(existing_vocabulary_size * 0.5)
+
+    def _add_buffer_to_vocabulary(self, attribute: Text) -> None:
+        """Add extra tokens to vocabulary for incremental training.
+
+        These extra tokens act as buffer slots which are used up sequentially
+        when more data is received as part of incremental training. Each of
+        these tokens start with a prefix `buf_` followed by the extra slot index.
+        So for example - buf_1, buf_2, buf_3... and so on.
+
+        Args:
+            attribute: Name of the attribute for which the vocabulary should be expanded.
+
+        """
+        original_vocabulary = self.vectorizers[attribute].vocabulary_
+        current_vocabulary_size = self._get_starting_empty_index(original_vocabulary)
```

Comment:
Should we simply use `len(original_vocabulary)` instead of iterating over all keys? We know that there is no buffer index, don't we?